Stage 3 - Implement Offline Capabilities with Cache

filipo203 opened the pull request:
Room database has been implemented into the app. In this version, data is fetched from the Dog API and saved into the database. The app uses said data from the database and uses it instead of calling straight from the API source.

Things that need checking:
- When attempting to view a dog breed that hasn't been downloaded to the database, a visual bug appears where all previously saved dog pictures from the database get displayed at random, when other images are being saved to the database until pictures are fetched. This can also cause the list of dog pictures to display incorrect breed pictures once pictures are fetched.
- App seems to have an issue displaying dog images when offline. Instead it displays blank placeholders but when clicked, app recognises image url but cannot display it as an image on DogPicturesScreen or DogPreviewScreen.

filipo203 commented:
COMMIT UPDATE:
- In database update 2, the method of fetching dog pictures from the API was changed. Dog pictures get downloaded for each breed when the dog breed gets selected from the UI's lazy list instead of all the images being fetched in the background all at once.

charlee-dev commented on app/src/main/java/com/example/dogadoption/DogApp.kt:
```diff
@@ -1,8 +1,24 @@
 
 @HiltAndroidApp
 class DogApp: Application() {
+/*
+    lateinit var database: DogDatabase
+
+    override fun onCreate() {
+        super.onCreate()
+        database = Room.databaseBuilder(
+            applicationContext,
+            DogDatabase::class.java,
+            "dog-database"
+        ).build()
+    }
+
+ */
+
```
leftover?

charlee-dev commented on app/src/main/java/com/example/dogadoption/repository/DogRepository.kt:
```diff
@@ -4,41 +4,67 @@ import android.content.ContentValues
+    suspend fun fetchAndSaveDogPictures(breed: String) {
+        val dogImageDB = localSource.getDogBreedImages(breed).firstOrNull()
+        if (dogImageDB.isNullOrEmpty()) {
+            try {
+                val result = remoteSource.getDogPictures(breed)
+                result.onSuccess { dogPictures ->
+                    dogPictures.forEach { imageUrl ->
+                        val existingImages = localSource.getDogBreedImages(breed).firstOrNull()
+                        if (existingImages == null || !existingImages.contains(imageUrl)) {
+                            val dogImages = DogImages(0, imageUrl, breed)
+                            localSource.insertDogBreedImages(dogImages)
+                        }
+                    }
+                    Log.d("DOG_REPOSITORY", "Dog pictures saved to database for $breed")
+                }.onFailure { e ->
+                    Log.e("DOG_REPOSITORY", "Failed to fetch dog pictures: ${e.message}")
+                    throw e
+                }
+            } catch (e: Exception) {
+                Log.e("DOG_REPOSITORY", "Failed to save dog pictures to database: ${e.message}")
+                throw e
```
1. This is a bit messy. You have a Kotlin result, which you unwrap. On failure, you throw, then catch again, and throw it again.
2. I would think that `localSource.getDogBreedImages(breed)` should never be null and return a list of images or an empty list
3. Every function should have a single responsibility so you know what it does when you read its name. Here, the name tells you that you do more than one thing. You can split this into three functions (or 4 if you put `instert` into separate function). 1st to get Images from db, 2nd to fetch images from remote, 3rd get images which would use 1st then if its empty try to pull 2nd, if it fails than it should return failure. I recommend you get familiar with [S.O.L.I.D. principles](https://www.youtube.com/watch?v=t8VTLxMsufU)

filipo203 commented:
COMMIT UPDATE (database update 3): 

- Renamed functions to indicate each function having a single responsibility.
- Organised code inside DogRepository for fun saveDogPictures (previously fetchAndSaveDogPictures).
- Fixed visual bug where dogs would show in lazy list at random by displaying circular progress indicator on screen when pictures are downloading to the database.
- Issue with images not displaying offline fixed by showing error message when no dog pictures from one breed are saved.
- Error message now appears where some dog pictures are missing from the lazy list when they should be displayed from the database.
- Adopt button won't appear on preview page when image isn't on screen.
- Removed some unnecessary code ie: some commented code, unused imports.
